### PR TITLE
introduce datasource resource

### PR DESCRIPTION
**Pull request by tdraier**
## Description

Use resource for datasource

## Risk

Something broken in datasource operations/usage. Can be rolledback if no further changes are done on top of it. 

## Deploy Plan

deploy `front`

### CODE REVIEW

File: front/lib/api/assistant/configuration.ts [hunk 1]
```diff
@@ -1490,15 +1491,10 @@ async function _createAgentDataSourcesConfigData(
...
-        where: {
-          workspaceId,
-          name: {
-            [Op.in]: dataSourceNames,
-          },
-        },
-        transaction: t,
-      });
+      return DataSourceResource.listByWorkspaceIdAndNames(
+        workspaceId,
```
Comment (flvndvd) [inline front/lib/api/assistant/configuration.ts:1495]: Either we pass `Authenticator` if we manage to remove the "multiple workspace ids" or we pass a `LightWorkspaceType` but no ModelId.

File: front/lib/documents_post_process_hooks/hooks/document_tracker/update_tracked_documents/lib.ts [hunk 2]
```diff
@@ -37,7 +37,7 @@ export async function shouldDocumentTrackerUpdateTrackedDocumentsRun(
     return false;
   }
 
-  const dataSource = await getDatasource(owner, params.dataSourceName);
+  const dataSource = await getDatasource(params.auth, params.dataSourceName);
```
Comment (flvndvd) [inline front/lib/documents_post_process_hooks/hooks/document_tracker/update_tracked_documents/lib.ts:40]: Can we pass params as a distinct parameters?

File: front/lib/resources/datasource_resource.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    });
+
+    if (!datasource) {
+      return null;
+    }
+    return new this(
+      DataSource,
+      datasource.get(),
+      datasource.editedByUser?.get()
+    );
```
Comment (flvndvd) [inline front/lib/resources/datasource_resource.ts:87]:
```suggestion
    return new this(
      DataSourceResource.model,
      datasource.get(),
      datasource.editedByUser?.get()
    );
```

File: front/lib/resources/string_ids.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@ const RESOURCES_PREFIX = {
   file: "fil",
   group: "grp",
   vault: "vlt",
+  datasource: "ds",
```
Comment (flvndvd) [inline front/lib/resources/string_ids.ts:25]: `dts` to keep the 3 letters prefix?

File: front/lib/workspace_usage.ts [hunk 5]
```diff
@@ -473,9 +474,9 @@ export async function checkWorkspaceActivity(workspace: Workspace) {
   const sevenDaysAgo = new Date();
   sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);
 
-  const hasDataSource = await DataSource.findOne({
-    where: { workspaceId: workspace.id },
-  });
+  const hasDataSource = await DataSourceResource.workspaceHasDatasources(
```
Comment (flvndvd) [inline front/lib/workspace_usage.ts:477]: This should probably live at the workspace level, instead can we do a regular `DataSourceResource.listXX()` for now and check that it returns more than 1 row? Maybe add a limit to this function.

File: front/pages/api/registry/[type]/lookup.ts [hunk 6]
```diff
@@ -87,12 +88,12 @@ async function handler(
...
           }
 
-          const dataSource = await DataSource.findOne({
-            where: {
-              workspaceId: owner.id,
-              name: req.query.data_source_id,
-            },
-          });
+          const dataSource = await DataSourceResource.fetchByName(
+            await Authenticator.internalBuilderForWorkspace(
```
Comment (flvndvd) [inline front/pages/api/registry/[type]/lookup.ts:92]: Can we make the authenticator outside of this 🙏?

File: front/lib/resources/datasource_resource.ts [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+      limit: 1,
+    });
+    return dataSources.length > 0;
+  }
+
+  async delete(
+    auth: Authenticator,
+    transaction?: Transaction
+  ): Promise<Result<undefined, Error>> {
+    await DataSourceViewResource.deleteForDataSource(auth, this.toJSON());
```
Comment (flvndvd) [inline front/lib/resources/datasource_resource.ts:160]: We should now pass the resource.

**Review by flvndvd — COMMENTED** [review]
This looks great 🔥

File: front/lib/resources/datasource_resource.ts [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            limit: options.limit,
+          }
+        : {}),
+      ...(options?.order
+        ? {
+            order: options.order,
+          }
+        : {}),
+    };
+  }
```
Comment (flvndvd) [inline front/lib/resources/datasource_resource.ts:74]: `nit`
```suggestion
private static getOptions(options?: FetchDataSourceOptions) {
  const result: FindOptions<this.model> = {};

  if (options?.includeEditedBy) {
    result.include = [
      {
        model: User,
        as: "editedByUser",
      },
    ];
  }

  if (options?.limit) {
    result.limit = options.limit;
  }

  if (options?.order) {
    result.order = options.order;
  }

  return result;
}
```

File: front/lib/resources/datasource_resource.ts [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+  static async fetchByName(
+    auth: Authenticator,
+    name: string,
+    options?: FetchDataSourceOptions
```
Comment (flvndvd) [inline front/lib/resources/datasource_resource.ts:79]: This method shoudn't accept limit or order, right?

File: front/pages/api/registry/[type]/lookup.ts [hunk 10]
```diff
@@ -86,13 +87,13 @@ async function handler(
...
             return;
           }
-
-          const dataSource = await DataSource.findOne({
-            where: {
-              workspaceId: owner.id,
-              name: req.query.data_source_id,
-            },
-          });
+          const auth = await Authenticator.internalBuilderForWorkspace(
```
Comment (flvndvd) [inline front/pages/api/registry/[type]/lookup.ts:90]: Here given that it's an API route we should be able to get an authenticator from the key

File: front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/linked_slack_channels.ts [hunk 11]
```diff
@@ -41,12 +39,11 @@ async function handler(
...
-  const slackDataSource = await DataSource.findOne({
-    where: {
-      connectorProvider: "slack",
-      workspaceId: owner.id,
-    },
-  });
+  const slackDataSources = await DataSourceResource.listByConnectorProvider(
+    auth,
+    "slack"
+  );
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/linked_slack_channels.ts:45]: `nit`
```suggestion
  const slackDataSources = await DataSourceResource.listByConnectorProvider(
    auth,
    "slack",
    {limit: 1}
  );
```

File: front/pages/api/w/[wId]/data_sources/[name]/index.ts [hunk 12]
```diff
@@ -42,8 +42,8 @@ async function handler(
     });
   }
 
-  const dataSource = await getDataSource(auth, req.query.name);
-  if (!dataSource) {
+  const dataSource = await DataSourceResource.fetchByName(auth, req.query.name);
+  if (!dataSource || !auth.isUser()) {
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/data_sources/[name]/index.ts:46]: The `auth.isUser()` is already above.

File: front/pages/api/w/[wId]/data_sources/[name]/index.ts [hunk 13]
```diff
@@ -141,20 +120,11 @@ async function handler(
           toUpdate.assistantDefaultSelected = req.body.assistantDefaultSelected;
         }
 
-        ds = await dataSourceModel.update(toUpdate);
+        await dataSourceResource.update(toUpdate);
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/data_sources/[name]/index.ts:123]: Shouldn't it be `dataSource` here?

File: front/pages/api/registry/[type]/lookup.ts (hunk 10 above)
Comment (tdraier) [reply front/pages/api/registry/[type]/lookup.ts:90]: no key for registry calls from the core (only a special secret)

File: front/lib/resources/datasource_resource.ts [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+        transaction,
+      });
+
+      return new Ok(undefined);
+    } catch (err) {
+      return new Err(err as Error);
+    }
+  }
+
+  async update(
```
Comment (flvndvd) [inline front/lib/resources/datasource_resource.ts:182]: It's okay for the refactoring to have a global `update`, let's make sure that we refine it as we implement group and vault, to have dedicated method for each update case.

**Review by flvndvd — APPROVED** [review]

**Pull request merged**